### PR TITLE
Using getent to access user and group databases

### PR DESCRIPTION
**Pull request by dipankar**
Using getent, and then falling back to files. Should help with #232

### CODE REVIEW

**Comment by vmarmol** [comment]
@dipankar you'll need to sign your commits for the CI to be happy.

**Comment by dipankar** [comment]
Fixed :+1:

File: user/user.go [hunk 1]
```diff
@@ -197,23 +199,36 @@ type ExecUser struct {
...
-	passwd, err := os.Open(passwdPath)
-	if err != nil {
-		passwd = nil
-	} else {
-		defer passwd.Close()
+
+	cmd := exec.Command("getent", "passwd")
+	stdout, err := cmd.Output()
+	passwd := bytes.NewReader(stdout)
+	if err == nil {
```
Comment (vmarmol) [inline user/user.go]: != nil?

Comment (dipankar) [inline user/user.go]: Yea, the conditions were incorrect! Fixed and squashed the commits.

**Comment by vmarmol** [comment]
Looks good overall, but @tianon and @cyphar know best :)

**Comment by mrunalp** [comment]
LGTM

**Comment by LK4D4** [comment]
@dipankar But this needs rebase :)

**Comment by dipankar** [comment]
@LK4D4 Done!

**Comment by LK4D4** [comment]
I wanted to write comment about `bytes.Buffer` as `cmd.Stdout`, but then realized this will be the same by memory consumption. Also maybe this can be extended to `key` parameter to getent later.
LGTM

**Comment by tianon** [comment]
Isn't this somewhat dangerous, or at the very least unexpected for users?

The reason I ask is that the first binary named "getent" in the container's
PATH will now be invoked inside the container before what the user
specified, which IMO is surprising behavior.

So, as much as I like the idea here, I'm -1 without more convincing that
this won't be a problem or cause unexpected side effects.

**Comment by cyphar** [comment]
**NOT** LGTM. This is wrong place to change the lookup code for passwd-formatted data. The correct place would be in `user/lookup_unix.go` -- and you would have to change both the `io.ReadCloser` and path interfaces (which would require temporary files, which I don't think is a good idea for a general purpose library call).

Also, I don't see what the benefit of this is. The point of `libcontainer/user` is to provide a pure Go replaceemnt for the `os/user` stdlib (which already uses the C stdlib). By executing `getent` (which is not safe, btw) you are just doing the same thing through the `getent` binary.

**Comment by dqminh** [comment]
i agree with @tianon here. Depending on the external binary `getent` and its existence seems weird.

Does it make sense if we write a wrapper to use the underlying functions directly i.e `setpwent / getpwent / endpwent` etc. ?

**Comment by cyphar** [comment]
@dqminh See my comment above about generic design. If you use the C functions, you can't parse container `/etc/passwd` and `/etc/group` data without being in the container. However, yours is a better solution IMO than executing a random command.

File: user/user.go [hunk 2]
```diff
@@ -197,23 +199,36 @@ type ExecUser struct {
 	Home     string
 }
 
-// GetExecUserPath is a wrapper for GetExecUser. It reads data from each of the
+// GetExecUserFile is a wrapper for GetExecUser.
```
Comment (cyphar) [inline user/user.go]: The function name is `GetExecUserPath`?

**Comment by dipankar** [comment]
@cyphar @tianon I do understand that executing getent make this sort of pointless, but as pointed out in #232 the existing solution does not get information from network sources. I was planning on initially abstracting it out to lookup_unix.go, but that would have changed quite a few things as you have pointed out. Thanks for the feedback! Will come up with a better solution.

**Comment by dqminh** [comment]
@cyphar @tianon actually im not sure of why cant we use `os/user` to lookup username/id here. iirc, when user is setup, we should already be inside the container so `os/user` should lookup correctly. Then we dont have to worry about this because `os/user` does a proper lookup in the passwd database.

File: user/user.go (hunk 2 above)
Comment (dipankar) [inline user/user.go]: Updated the comment

**Comment by cyphar** [comment]
@dqminh The reason why the library looks the way it does is because it was designed with the goal of processing container passwd-formatted data from the host (why? initially in preperation for docker#9046, but also because it makes it a "complete" solution). Again, you are free to add some extra `GetPasswdEnt` or whatever functions that return `io.Reader`s which you can then pass to the general `io.Reader` `GetExecUser` (that is the **entire reason** that it exists, is to solve this problem).

**Comment by tianon** [comment]
Also, to use "os/user" we have to link against nss (IIRC), which is why
that stdlib can't work in the general case.

**Pull request closed without merging**